server.go

```diff
@@ -302,14 +328,35 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
 		if e == p {
 			// Issue an asynchronous reconnect if the peer was a
 			// persistent outbound connection.
-			if !p.inbound && p.persistent && atomic.LoadInt32(&s.shutdown) == 0 {
+			peerCfg := &peer.Config{
+				Proxy:            cfg.Proxy,
+				SimNet:           cfg.SimNet,
+				RegressionTest:   cfg.RegressionTest,
+				DisableListen:    cfg.DisableListen,
+				Oniondial:        cfg.oniondial,
+				Dial:             cfg.dial,
+				UserAgentName:    userAgentName,
+				UserAgentVersion: userAgentVersion,
+			}
+			if !p.Inbound() && p.Persistent() && atomic.LoadInt32(&s.shutdown) == 0 {
 				delete(list, e)
-				e = newOutboundPeer(s, p.addr, true, p.retryCount+1)
+				e = peer.NewOutboundPeer(s.db.NewestSha, s.nonce, s.donePeers,
+					s.chainParams.Net, s.addrManager, peerCfg, p.Addr(), true, p.RetryCount()+1)
 				list[e] = struct{}{}
+				handleVersionMsg := func(p *peer.Peer, msg *wire.MsgVersion) {
+					// Add the remote peer time as a sample for creating an offset against
+					// the local clock to keep the network time in sync.
+					s.timeSource.AddTimeSample(p.Addr(), msg.Timestamp)
+
+					// Signal the block manager this peer is a new sync candidate.
+					s.blockManager.NewPeer(p)
+				}
+				p.AddVersionMsgListener("handleVersionMsg", handleVersionMsg)
```

Comment:
Isn't this (and the code before it) already done by the call to `s.registerListeners(p)` below?  If not, they're using the same key so it would be overwritten anyways.